File: JANUARY/09-01-2026/865.php
class Solution {
    function subtreeWithAllDeepest($root) {
        $dfs = function($node) use (&$dfs) {
            if ($node === null) return [0, null];
            [$ld, $ln] = $dfs($node->left);
            [$rd, $rn] = $dfs($node->right);
            if ($ld > $rd) return [$ld + 1, $ln];
            if ($rd > $ld) return [$rd + 1, $rn];
            return [$ld + 1, $node];
        };
        return $dfs($root)[1];
    }
}

File: JANUARY/09-01-2026/865.rs
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn subtree_with_all_deepest(
        root: Option<Rc<RefCell<TreeNode>>>
    ) -> Option<Rc<RefCell<TreeNode>>> {
        fn dfs(
            node: Option<Rc<RefCell<TreeNode>>>
        ) -> (i32, Option<Rc<RefCell<TreeNode>>>) {
            match node {
                None => (0, None),
                Some(n) => {
                    let left = n.borrow().left.clone();
                    let right = n.borrow().right.clone();
                    let (ld, ln) = dfs(left);
                    let (rd, rn) = dfs(right);
                    if ld > rd {
                        (ld + 1, ln)
                    } else if rd > ld {
                        (rd + 1, rn)
                    } else {
                        (ld + 1, Some(n))
                    }
                }
            }
        }
        dfs(root).1
    }
}